libmamba/include/mamba/core/solver.hpp

```diff
@@ -68,20 +68,21 @@ namespace mamba
         void add_pins(const std::vector<std::string>& pins);
         void set_flags(const std::vector<std::pair<int, int>>& flags);
         void set_postsolve_flags(const std::vector<std::pair<int, int>>& flags);
-        [[nodiscard]] bool is_solved() const;
         [[nodiscard]] bool try_solve();
         void must_solve();
-        std::string problems_to_str() const;
-        std::vector<std::string> all_problems() const;
-        std::vector<MSolverProblem> all_problems_structured() const;
-        std::string all_problems_to_str() const;
+
+        [[nodiscard]] bool is_solved() const;
```

Comment:
I am not sure if the `nodiscard` really does something useful here besides adding noise to the header. I guess the only place where I think it makes a lot of sense is the `solve()` method (because we want to enforce users to check that it was actually solved.

Or am I misunderstanding `nodiscard`?

Reply:
I have the same understanding as you. On `solve` it is critical (as it happens, it is currently *not* checked everywhere).

It is however correct to use it here, because there is no point in calling `is_solved` without checking the result because this function has no side effect. Some best-practices suggest to use it liberally in those case, it's only more bugs you can catch (*e.g.* if the `const` function name is ambiguous). There's even a `clang-tidy` fix that would apply it everywhere it can.

Reply:
I agree that `[[nodiscard]]` seems useful here, at least to catch cases where it's called but not checked.